lib/metrics.js

```diff
@@ -0,0 +1,129 @@
+    this.events.subscribe('before-keyboard-navigate', this.onKeyNavigate);
+    this.events.subscribe('before-click-navigate', this.onClickNavigate);
+
+    // Reset the interaction state when the urlbar contents change.
+    this.events.subscribe('urlbar-change', this.interactionStart);
+
+    // The interaction ends when the popup closes.
+    this.events.subscribe('before-popup-hide', this.interactionEnd);
+  },
+  destroy: function() {
+    this.win.console.log('metrics.destroy');
+    this.events.unsubscribe('before-popup-show', this.interactionStart);
+    this.events.unsubscribe('before-keyboard-navigate', this.onKeyNavigate);
+    this.events.unsubscribe('before-click-navigate', this.onClickNavigate);
+    this.events.unsubscribe('urlbar-change', this.interactionStart);
+    this.events.unsubscribe('before-popup-hide', this.interactionEnd);
+  },
+  interactionStart: function() {
+    this.win.console.log('metrics.interactionStart');
+    // Initialize default values for the case where the popup closes without a
+    // navigation.
+    this.checked = false;
+    this.interaction = {
+
+      // Did the user navigate?
+      didNavigate: false,
+
+      // Was the selected item clicked or keyed or neither?
+      interactionType: null,
+
+      // Was the recommendation shown?
+      recommendationShown: false,
+
+      // What type of recommendation was it?
+      recommendationType: null,
+
+      // Was the recommendation the selected item?
+      recommendationSelected: null,
+
+      // What was the index of the selected item?
+      // If the recommendation was selected, this will be -1.
+      selectedIndex: null
+    };
+  },
+  onKeyNavigate: function() {
+    this.win.console.log('metrics.onKeyNavigate');
+    this.checkState('key');
+  },
+  onClickNavigate: function(data) {
+    this.win.console.log('metrics.onClickNavigate');
+    // It turns out that the popup's click handler can get the selectedIndex
+    // before the popup starts to handle it, but checkState is too late, and
+    // always thinks selectedIndex = -1 in the case of clicks.
+    this.win.console.log('onClickNavigate input: ', JSON.stringify(data));
+    this.checkState('click', data.selectedIndex);
+  },
+  checkState: function(interactionType, selectedIndex) {
+    this.win.console.log('metrics.checkState, selectedIndex is ', selectedIndex);
+    this.checked = true;
+    this.interaction = {
+      // If interactionType is defined, then we did navigate.
+      didNavigate: !!interactionType,
+      interactionType: interactionType || null,
+      recommendationShown: !this.recommendation.el.collapsed,
+      recommendationSelected: this.recommendation.isHighlighted,
+      recommendationType: this.recommendation.resultType,
+      // If no selectedIndex was passed in, we get the value from the popup.
+      // Note that `isNaN(undefined) => true`.
+      selectedIndex: isNaN(selectedIndex) ? this.popup.el.selectedIndex : selectedIndex
```

Comment:
Hm. It's possible that the user might hit a letter and the enter key faster than the popup can open for the first time. In that case, `this.popup.el` might not exist yet, so that this would throw.